nemo/collections/tts/modules/fastpitch.py

```diff
@@ -230,19 +236,27 @@ def forward(
 
         # Input FFT
         enc_out, enc_mask = self.encoder(input=text, conditioning=spk_emb)
-
-        log_durs_predicted = self.duration_predictor(enc_out, enc_mask)
+        if self.speaker_emb_condition_prosody:
+            log_durs_predicted = self.duration_predictor(enc_out + spk_emb, enc_mask)
+        else:
+            log_durs_predicted = self.duration_predictor(enc_out, enc_mask) 
         durs_predicted = torch.clamp(torch.exp(log_durs_predicted) - 1, 0, self.max_token_duration)
 
         attn_soft, attn_hard, attn_hard_dur, attn_logprob = None, None, None, None
         if self.learn_alignment and spec is not None:
             text_emb = self.encoder.word_emb(text)
-            attn_soft, attn_logprob = self.aligner(spec, text_emb.permute(0, 2, 1), enc_mask == 0, attn_prior)
+            if self.speaker_emb_condition_aligner and not isinstance(spk_emb, int):
+                attn_soft, attn_logprob = self.aligner(spec, text_emb.permute(0, 2, 1), enc_mask == 0, attn_prior, conditioning=spk_emb)
+            else:
+                attn_soft, attn_logprob = self.aligner(spec, text_emb.permute(0, 2, 1), enc_mask == 0, attn_prior)
             attn_hard = binarize_attention_parallel(attn_soft, input_lens, mel_lens)
             attn_hard_dur = attn_hard.sum(2)[:, 0, :]
 
         # Predict pitch
-        pitch_predicted = self.pitch_predictor(enc_out, enc_mask)
+        if self.speaker_emb_condition_prosody:
+            pitch_predicted = self.pitch_predictor(enc_out + spk_emb, enc_mask)
```

Comment:
Summing features/embeddings is usually only done with transformers because of the limitation that their input & output dimensions must be equal. For other models like DNN/RNN/CNN the standard is to concatenate the feature vectors together. The latter approach looks like what is done in the RadTTS code as well. Did we ever try it?

To save refactoring time, it may be worth changing these booleans to string/enum and having something like:

```
if not self.speaker_emb_condition_prosody:
  prosody_input = enc_out
elif self.speaker_emb_condition_prosody == "add":
  prosody_input = enc_out + spk_emb
elif self.speaker_emb_condition_prosody == "concatenate":
  spk_emb_tile = spk_emb.unsqueeze(1).tile([1, text.shape[1], 1])
  prosody_input = torch.concat([enc_out, spk_emb_tile], dim=2)
else:
  raise ValueError()
```

Reply:
Cheng-Ping had experimented with this and we did not get very good results or better results than what we have now. We can definitely add this feature in the future but it would need more experimentation.

Reply:
My concern would be leaving it as a boolean means that fixing it later will break backwards compatibility. And adding in the functionality seems inevitable because:

(1) RadTTS does it, and we will need to merge/consolidate the common components.
(2) Concatenation creates a non-linearity with the feature vector that is needed if you condition on multiple feature vectors at once, or use a static fixed external feature vector like from eg. a speaker embedding or language model.

Though I guess (1) will be backwards incompatible regardless, so I guess we could leave it for then.